aws/resource_aws_emr_cluster.go

```diff
@@ -1741,3 +1962,272 @@ func fetchAllEMRInstanceGroups(conn *emr.EMR, clusterID string) ([]*emr.Instance
+	}
+	if spotSpecification.BlockDurationMinutes != nil {
+		m["block_duration_minutes"] = aws.Int64Value(spotSpecification.BlockDurationMinutes)
+	}
+	if spotSpecification.AllocationStrategy != nil {
+		m["allocation_strategy"] = aws.StringValue(spotSpecification.AllocationStrategy)
+	}
+
+	return []interface{}{m}
+}
+
+func expandEbsConfiguration(ebsConfigurations []interface{}) *emr.EbsConfiguration {
+	ebsConfig := &emr.EbsConfiguration{}
+	ebsConfigs := make([]*emr.EbsBlockDeviceConfig, 0)
+	for _, ebsConfiguration := range ebsConfigurations {
+		cfg := ebsConfiguration.(map[string]interface{})
+		ebsBlockDeviceConfig := &emr.EbsBlockDeviceConfig{
+			VolumesPerInstance: aws.Int64(int64(cfg["volumes_per_instance"].(int))),
+			VolumeSpecification: &emr.VolumeSpecification{
+				SizeInGB:   aws.Int64(int64(cfg["size"].(int))),
+				VolumeType: aws.String(cfg["type"].(string)),
+			},
+		}
+		if v, ok := cfg["iops"].(int); ok && v != 0 {
+			ebsBlockDeviceConfig.VolumeSpecification.Iops = aws.Int64(int64(v))
+		}
+		ebsConfigs = append(ebsConfigs, ebsBlockDeviceConfig)
+	}
+	ebsConfig.EbsBlockDeviceConfigs = ebsConfigs
+	return ebsConfig
+}
+
+func expandInstanceTypeConfigs(instanceTypeConfigs []interface{}) []*emr.InstanceTypeConfig {
+	configsOut := []*emr.InstanceTypeConfig{}
+
+	for _, raw := range instanceTypeConfigs {
+		configAttributes := raw.(map[string]interface{})
+
+		config := &emr.InstanceTypeConfig{
+			InstanceType: aws.String(configAttributes["instance_type"].(string)),
+		}
+
+		if bidPrice, ok := configAttributes["bid_price"]; ok {
+			if bidPrice != "" {
+				config.BidPrice = aws.String(bidPrice.(string))
+			}
+		}
+
+		if v, ok := configAttributes["bid_price_as_percentage_of_on_demand_price"].(float64); ok && v != 0 {
+			config.BidPriceAsPercentageOfOnDemandPrice = aws.Float64(v)
+		}
+
+		if v, ok := configAttributes["weighted_capacity"].(int); ok {
+			config.WeightedCapacity = aws.Int64(int64(v))
+		}
+
+		if v, ok := configAttributes["configurations"].(*schema.Set); ok && v.Len() > 0 {
+			config.Configurations = expandConfigurations(v.List())
+		}
+
+		if v, ok := configAttributes["ebs_config"].(*schema.Set); ok && v.Len() == 1 {
+			config.EbsConfiguration = expandEbsConfiguration(v.List())
+		}
+
+		configsOut = append(configsOut, config)
+	}
+
+	return configsOut
+}
+
+func expandLaunchSpecification(launchSpecification map[string]interface{}) *emr.InstanceFleetProvisioningSpecifications {
+	onDemandSpecification := launchSpecification["on_demand_specification"].([]interface{})
+	spotSpecification := launchSpecification["spot_specification"].([]interface{})
+
+	fleetSpecification := &emr.InstanceFleetProvisioningSpecifications{}
+
+	if len(onDemandSpecification) > 0 {
+		fleetSpecification.OnDemandSpecification = &emr.OnDemandProvisioningSpecification{
+			AllocationStrategy: aws.String(onDemandSpecification[0].(map[string]interface{})["allocation_strategy"].(string)),
+		}
+	}
+
+	if len(spotSpecification) > 0 {
```

Comment:
To prevent potential panics, we should guard the type assertions and include additional `nil` checks here:

```suggestion
	fleetSpecification := &emr.InstanceFleetProvisioningSpecifications{}

	if onDemandSpecification, ok := launchSpecification["on_demand_specification"].([]interface{}); ok && len(onDemandSpecification) > 0 && onDemandSpecification[0] != nil {
		fleetSpecification.OnDemandSpecification = &emr.OnDemandProvisioningSpecification{
			AllocationStrategy: aws.String(onDemandSpecification[0].(map[string]interface{})["allocation_strategy"].(string)),
		}
	}

	if spotSpecification, ok := launchSpecification["spot_specification"].([]interface{}); ok && len(spotSpecification) > 0 && spotSpecification[0] != nil {
```